### PR TITLE
[VL] Optimize GHA workflow with nightly built docker image (part-1)

### PR DESCRIPTION
**Pull request by PZD-CHINA**
The patch targets to do CI cleanup.

.github/workflows/velox_be.yml static-build-test and  build-script-test need further debug.

### CODE REVIEW

File: .github/workflows/velox_be.yml [hunk 1]
```diff
@@ -47,82 +47,91 @@ concurrency:
   group: ${{ github.repository }}-${{ github.head_ref || github.sha }}-${{ github.workflow }}
   cancel-in-progress: true
 
-jobs:
+env:
+  HTTP_PROXY_HOST: 10.240.252.16
+  HTTP_PROXY_PORT: 911
+  DOCKER_REGISTRY_HOST: 10.0.0.25
+  DOCKER_REGISTRY_PORT: 5000
+  PATH_TO_GLUTEN_TE: ./tools/gluten-te
```
Comment (PHILO-HE) [inline .github/workflows/velox_be.yml:55]: This part for defining env can be unified with the above.

File: tools/gluten-te/centos/buildenv.sh [hunk 2]
```diff
@@ -42,14 +42,12 @@ TIMEZONE=${TIMEZONE:-$DEFAULT_TIMEZONE}
 
 # Set operating system
 OS_IMAGE=${OS_IMAGE:-$DEFAULT_OS_IMAGE}
-
-# Set os version
```
Comment (PHILO-HE) [inline tools/gluten-te/centos/buildenv.sh:46]: removed unintentionally?

File: tools/gluten-te/centos/defaults.conf [hunk 3]
```diff
@@ -21,17 +21,28 @@ DEFAULT_DEBUG_BUILD=OFF
 DEFAULT_BUILD_BACKEND_TYPE=velox
 
 # HTTP proxy
-DEFAULT_HTTP_PROXY_HOST=
-DEFAULT_HTTP_PROXY_PORT=
+# If http proxy is http://child-prc.intel.com:913, write it as
+# DEFAULT_HTTP_PROXY_HOST=child-prc.intel.com
```
Comment (PHILO-HE) [inline tools/gluten-te/centos/defaults.conf:25]: Since `HTTP_PROXY_HOST & HTTP_PROXY_PORT` are already defined at the beginning of velox_be.yml, it seems there is no need to set the default here. Ditto for ubuntu/defaults.conf.

File: .github/workflows/velox_be.yml [hunk 4]
```diff
@@ -413,49 +433,52 @@ jobs:
       - name: Exit docker container
         if: ${{ always() }}
         run: |
-          docker stop centos8-test-$GITHUB_RUN_ID || true
-
+          $PATH_TO_GLUTEN_TE/$OS_IMAGE/gha/gha-checkout/clean.sh
+
```
Comment (PHILO-HE) [inline .github/workflows/velox_be.yml:437]: Nit: remove space character. Still keep an empty line.

**Comment by PHILO-HE** [comment]
Please replace `/opt/spark331` by `/opt/spark342` in the below two pieces of code.
Using spark 331 binary for spark 342 tests has some potential issues. In the new docker, suppose spark-342 source is downloaded and then built. See [code link](https://github.com/oap-project/gluten/blob/main/tools/gluten-te/ubuntu/dockerfile-buildenv#L114).

https://github.com/oap-project/gluten/blob/29f850086609d8922b30c163ea1687fbd25a2e8d/.github/workflows/velox_be.yml#L238

https://github.com/oap-project/gluten/blob/29f850086609d8922b30c163ea1687fbd25a2e8d/.github/workflows/velox_be.yml#L276

File: .github/workflows/velox_be.yml [hunk 5]
```diff
@@ -40,89 +40,91 @@ on:
 env:
   HTTP_PROXY_HOST: 10.240.252.16
   HTTP_PROXY_PORT: 911
-  DOCKER_REGISTRY_HOST: 10.0.2.4
+  DOCKER_REGISTRY_HOST: 10.0.0.25
   DOCKER_REGISTRY_PORT: 5000
-
+  PATH_TO_GLUTEN_TE: ./tools/gluten-te
```
Comment (zhouyuan) [inline .github/workflows/velox_be.yml:45]: I think we should be able to use the Github environment variables

Comment (PHILO-HE) [reply .github/workflows/velox_be.yml:45]: Hongze once removed the use of github environment in a revert commit: https://github.com/oap-project/gluten/commit/385b252468dddfce43ccd1071961e9deb71fbaa5. I will check this with him.

**Comment by zhouyuan** [comment]
please help to do a rebase, the CK backend failure is fixed in main branch

**Comment by JkSelf** [comment]
@PZD-CHINA I'm not sure why these changes caused the failure in spark 34. Could you please disable these two failed unit tests and added them in this list for spark34 failed unit tests https://github.com/oap-project/gluten/issues/3559? Thanks.

**Comment by zhztheplayer** [comment]
> @PZD-CHINA I'm not sure why these changes caused the failure in spark 34. Could you please disable these two failed unit tests and added them in this list for spark34 failed unit tests #3559? Thanks.

I see the patch corrects 3.4 UT's test property `spark.test.home`:

https://github.com/oap-project/gluten/pull/4293/files#diff-92077173e6a535009e50974639976b3c49d45cfb05141bac9d291fc13bb06a8aL223-L224

https://github.com/oap-project/gluten/pull/4293/files#diff-92077173e6a535009e50974639976b3c49d45cfb05141bac9d291fc13bb06a8aL257-L261

So the change could be related to the new failures?

**Comment by PHILO-HE** [comment]
> > @PZD-CHINA I'm not sure why these changes caused the failure in spark 34. Could you please disable these two failed unit tests and added them in this list for spark34 failed unit tests #3559? Thanks.
> 
> I see the patch corrects 3.4 UT's test property `spark.test.home`:
> 
> https://github.com/oap-project/gluten/pull/4293/files#diff-92077173e6a535009e50974639976b3c49d45cfb05141bac9d291fc13bb06a8aL223-L224
> 
> https://github.com/oap-project/gluten/pull/4293/files#diff-92077173e6a535009e50974639976b3c49d45cfb05141bac9d291fc13bb06a8aL257-L261
> 
> So the change could be related to the new failures?

I guess so. @PZD-CHINA, let's make spark 3.4 UT failure ignored in this PR.

File: tools/gluten-te/ubuntu/cbash-build.sh [hunk 6]
```diff
@@ -56,4 +57,4 @@ CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS $EXTRA_DOCKER_OPTIONS"
 CBASH_BASH_ARGS="$*"
 BASH_ARGS="$CBASH_BASH_ARGS"
 
-docker run $CBASH_DOCKER_RUN_ARGS $DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE "cd /opt/gluten && $BASH_ARGS"
+docker run $CBASH_DOCKER_RUN_ARGS $DOCKER_PUSH_REGISTRY/$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE "cd /opt/gluten && $BASH_ARGS"
```
Comment (zhztheplayer) [inline tools/gluten-te/ubuntu/cbash-build.sh:60]: Any reason making the change? Thanks!

**Review by zhztheplayer — COMMENTED** [review]
Thanks you guys for working on this!

File: .github/workflows/velox_be.yml [hunk 7]
```diff
@@ -40,89 +40,92 @@ on:
...
 
 jobs:
-
   ubuntu2004-test-spark32:
     runs-on: velox-self-hosted
+    env:
+      OS_IMAGE: ubuntu
+      OS_VERSION: 20.04
     steps:
       - uses: actions/checkout@v4
```
Comment (zhztheplayer) [inline .github/workflows/velox_be.yml:59]: `- uses: actions/checkout@v4`

I think this step can be removed since we are now checking out code inside the docker via `checkout.sh`.

File: tools/gluten-te/ubuntu/cbash-build.sh (hunk 6 above)
Comment (PZD-CHINA) [reply tools/gluten-te/ubuntu/cbash-build.sh:60]: The format for the images' name is $DOCKER_PUSH_REGISTRY/$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE, rather than $DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE. The prefix $DOCKER_PUSH_REGISTRY/ is necessary to correctly pull the images.

File: tools/gluten-te/centos/cbash-build.sh [hunk 8]
```diff
@@ -46,6 +46,7 @@ CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS --privileged"
 CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS --ulimit nofile=65536:65536"
 CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS --ulimit core=-1"
 CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS --security-opt seccomp=unconfined"
+CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS -e NUM_THREADS=30"
```
[the remark this answers is not in the document]
Comment (PZD-CHINA) [reply tools/gluten-te/centos/cbash-build.sh:49]: Would "NUM_THREADS=30" be suitable for all environments?
Yes. It has been tested and found to be applicable to all environments, and it aligns with the previous YAML file.

**Comment by zhouyuan** [comment]
it seems blow url for boost is invalid now
 https://boostorg.jfrog.io/artifactory/main/release/1.72.0/source/boost_1_72_0.tar.gz

File: tools/gluten-te/centos/cbash-build.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply tools/gluten-te/centos/cbash-build.sh:49]: Thanks for explaining.

I would assume that there is one case that a user could follow the workflow file to build Gluten on their PC which has limited cores (e.g. 4 cores or 8 cores). 

But it's not a very big problem and we can let the code as it is, then do some optimizations later on.

File: tools/gluten-te/ubuntu/cbash-build.sh (hunk 6 above)
Comment (zhztheplayer) [reply tools/gluten-te/ubuntu/cbash-build.sh:60]: Thanks and I understand your point. However I would be inclined to still use `$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE` rather than `$DOCKER_PUSH_REGISTRY/$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE` since the purpose of the code here was to locate and use the local image we just built within `build.sh`.

See code https://github.com/oap-project/gluten/blob/3fe4c959e07739ac142433485691f92a8a75bfd1/tools/gluten-te/ubuntu/build.sh#L73C51-L73C94

We used `-t $DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE` to create a local image named `$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE`, during which we use `$DOCKER_PUSH_REGISTRY/$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE` as cache through command `--cache-from $DOCKER_CACHE_REGISTRY/$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE`.

So the pipeline here is, use remote image as cache, to build the local image.

The pros of this way, is when `DOCKER_CACHE_REGISTRY` is not set, the workflow / user could still be able to perform build to generate a local image `$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE` as if encountering a cache miss.

As an experiment, you can check whether your new way in the patch could be able to handle the case that `DOCKER_CACHE_REGISTRY` is not set.

Comment (PZD-CHINA) [reply tools/gluten-te/ubuntu/cbash-build.sh]: You are right, there should not be any changes made here. It has been modified.

File: tools/gluten-te/ubuntu/dockerfile-buildenv [hunk 9]
```diff
@@ -1,6 +1,7 @@
 ARG OS_IMAGE
+ARG OS_VERSION
 
-FROM $OS_IMAGE AS gluten-buildenv
+FROM $OS_IMAGE:$OS_VERSION AS gluten-buildenv
```
[the remark this answers is not in the document]
Comment (PZD-CHINA) [reply tools/gluten-te/ubuntu/dockerfile-buildenv]: Yes. I standardized the name in new commit.

File: tools/gluten-te/ubuntu/tpc.sh [hunk 10]
```diff
@@ -75,7 +75,7 @@ then
   exit 1
 fi
 
-DOCKER_SELECTED_TARGET_IMAGE_TPC_WITH_OS_IMAGE="$DOCKER_SELECTED_TARGET_IMAGE_TPC-$OS_IMAGE"
+DOCKER_SELECTED_TARGET_IMAGE_TPC_WITH_OS_IMAGE="$DOCKER_SELECTED_TARGET_IMAGE_TPC-$OS_IMAGE_NAME"
```
Comment (zhztheplayer) [inline tools/gluten-te/ubuntu/tpc.sh:78]: We may do the same on tpc scripts to replaced `$OS_IMAGE` by `$OS_IMAGE:$OS_IMAGE_TAG`. But it's non-blocking and can be done in a separate ticket.

File: tools/gluten-te/centos/cbash-build.sh [hunk 11]
```diff
@@ -56,4 +57,4 @@ CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS $EXTRA_DOCKER_OPTIONS"
 CBASH_BASH_ARGS="$*"
 BASH_ARGS="$CBASH_BASH_ARGS"
 
-docker run $CBASH_DOCKER_RUN_ARGS $DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE bash -c "cd /opt/gluten && $BASH_ARGS"
+docker run $CBASH_DOCKER_RUN_ARGS $DOCKER_PUSH_REGISTRY/$DOCKER_TARGET_IMAGE_BUILD_WITH_OS_IMAGE bash -c "cd /opt/gluten && $BASH_ARGS"
```
Comment (zhztheplayer) [inline tools/gluten-te/centos/cbash-build.sh:60]: ditto on `$DOCKER_PUSH_REGISTRY/`

File: tools/gluten-te/centos/cbash-mount.sh [hunk 12]
```diff
@@ -57,4 +57,4 @@ CBASH_DOCKER_RUN_ARGS="$CBASH_DOCKER_RUN_ARGS $EXTRA_DOCKER_OPTIONS"
 CBASH_BASH_ARGS="$*"
 BASH_ARGS="$CBASH_BASH_ARGS"
 
-docker run $CBASH_DOCKER_RUN_ARGS $DOCKER_TARGET_IMAGE_BUILDENV_WITH_OS_IMAGE bash -c "cd /opt/gluten && $BASH_ARGS"
+docker run $CBASH_DOCKER_RUN_ARGS $DOCKER_PUSH_REGISTRY/$DOCKER_TARGET_IMAGE_BUILDENV_WITH_OS_IMAGE bash -c "cd /opt/gluten && $BASH_ARGS"
```
Comment (zhztheplayer) [inline tools/gluten-te/centos/cbash-mount.sh:60]: ditto on `$DOCKER_PUSH_REGISTRY/`

**Review by zhztheplayer — COMMENTED** [review]
Remaining some dittos from my side. Thanks!

File: tools/gluten-te/centos/defaults.conf [hunk 13]
```diff
@@ -21,17 +21,25 @@ DEFAULT_DEBUG_BUILD=OFF
...
 
 # Base operator system image used in build scripts.
-DEFAULT_OS_IMAGE=centos:8
+# DEFAULT_OS_IMAGE_NAME=ubuntu or centos
+DEFAULT_OS_IMAGE_NAME=
 
 # Version ID of os image
-DEFAULT_OS_VERSION=8
+# DEFAULT_OS_IMAGE_TAG=20.04 or 8
+DEFAULT_OS_IMAGE_TAG=
```
Comment (zhztheplayer) [inline tools/gluten-te/centos/defaults.conf:42]: Minor: Why removing the default values? We'd better to let user having a default settings to run the scripts. For centos scripts it's OK to have a default base image `centos:8` IIUC.

File: tools/gluten-te/centos/defaults.conf [hunk 14]
```diff
@@ -43,10 +51,14 @@ DEFAULT_DOCKER_TARGET_IMAGE_BUILDENV=gluten-te/gluten-buildenv
...
-DEFAULT_DOCKER_CACHE_REGISTRY=
+# DEFAULT_DOCKER_CACHE_REGISTRY=
+# for 10.0.0.25 test 
+DEFAULT_DOCKER_CACHE_REGISTRY=10.0.0.25:5000
 
 # Docker registry to push pre-built images
-DEFAULT_DOCKER_PUSH_REGISTRY=
+# DEFAULT_DOCKER_PUSH_REGISTRY=
+# for 10.0.0.25 test
+DEFAULT_DOCKER_PUSH_REGISTRY=10.0.0.25:5000
```
Comment (zhztheplayer) [inline tools/gluten-te/centos/defaults.conf:61]: We'd revert the change before merging? do I understand it correctly?

File: tools/gluten-te/centos/defaults.conf (hunk 13 above)
Comment (zhztheplayer) [inline tools/gluten-te/centos/defaults.conf:42]: Minor: Why removing the default values? We'd better to let user having a default settings to run the scripts. For centos scripts it's OK to have a default base image `centos:8` IIUC.

Comment (PZD-CHINA) [reply tools/gluten-te/centos/defaults.conf]: OK. I've already made the modifications to the defaults.conf. Now there is a default base image conf for both CentOS and Ubuntu.

File: tools/gluten-te/ubuntu/tpc.sh (hunk 10 above)
Comment (PZD-CHINA) [reply tools/gluten-te/ubuntu/tpc.sh]: Corrected. Still use $OS_IMAGE.

File: tools/gluten-te/centos/defaults.conf (hunk 14 above)
Comment (PZD-CHINA) [reply tools/gluten-te/centos/defaults.conf:59]: I want to keep the settings of DEFAULT_DOCKER_PUSH_REGISTRY and DEFAULT_DOCKER_CACHE_REGISTRY in this place. Are there any problems with this?

 If revert changes, where should these variables be set?

**Comment by zhztheplayer** [comment]
LGTM. Thanks!

**Review by zhztheplayer — APPROVED** [review]

File: tools/gluten-te/centos/cbash-build.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (PZD-CHINA) [reply tools/gluten-te/centos/cbash-build.sh]: I understand what you mean. The 'NUM_THREADS=30' set has been removed.

File: tools/gluten-te/centos/build.sh [hunk 15]
```diff
@@ -17,7 +17,7 @@
 set -ex
 
 BASEDIR=$(dirname $0)
-
+source "$BASEDIR/defaults.conf"
```
[the remark this answers is not in the document]
Comment (PZD-CHINA) [reply tools/gluten-te/centos/build.sh]: Thank you for the reminder. This line has been deleted。

File: tools/gluten-te/centos/defaults.conf (hunk 14 above)
Comment (PZD-CHINA) [reply tools/gluten-te/centos/defaults.conf:59]: I plan to revert the change in defaults.conf and set DOCKER_PUSH_REGISTRY and DOCKER_CACHE_REGISTRY in .github/workflow/velox_be.yml

**Comment by ayushi-agarwal** [comment]
> @PZD-CHINA I'm not sure why these changes caused the failure in spark 34. Could you please disable these two failed unit tests and added them in this list for spark34 failed unit tests #3559? Thanks.

@JkSelf This PR will solve the issue https://github.com/oap-project/gluten/issues/4203. As part of this PR https://github.com/oap-project/gluten/pull/4273 we found that these were failing in our local when we give correct path and many other UT's pass when given the correct path.

**Comment by PHILO-HE** [comment]
Let's leave the change for yml in a follow-up PR and then really use the new docker after that.

**Review by PHILO-HE — APPROVED** [review]
Thanks for your work!

**Comment by vibhaska** [comment]
> Please replace `/opt/spark331` by `/opt/spark342` in the below two pieces of code. Using spark 331 binary for spark 342 tests has some potential issues. In the new docker, suppose spark-342 source is downloaded and then built. See [code link](https://github.com/oap-project/gluten/blob/main/tools/gluten-te/ubuntu/dockerfile-buildenv#L114).
> 
> https://github.com/oap-project/gluten/blob/29f850086609d8922b30c163ea1687fbd25a2e8d/.github/workflows/velox_be.yml#L238
> 
> https://github.com/oap-project/gluten/blob/29f850086609d8922b30c163ea1687fbd25a2e8d/.github/workflows/velox_be.yml#L276

Has this been done already?

**Comment by vibhaska** [comment]
> > Please replace `/opt/spark331` by `/opt/spark342` in the below two pieces of code. Using spark 331 binary for spark 342 tests has some potential issues. In the new docker, suppose spark-342 source is downloaded and then built. See [code link](https://github.com/oap-project/gluten/blob/main/tools/gluten-te/ubuntu/dockerfile-buildenv#L114).
> > https://github.com/oap-project/gluten/blob/29f850086609d8922b30c163ea1687fbd25a2e8d/.github/workflows/velox_be.yml#L238
> > 
> > https://github.com/oap-project/gluten/blob/29f850086609d8922b30c163ea1687fbd25a2e8d/.github/workflows/velox_be.yml#L276
> 
> Has this been done already?

@JkSelf Possibly missed in checkin because its a ".github" folder, ignored by git.
